In [174]:
import pandas as pd

In [175]:
know_2020 = pd.read_csv('../know_Data/train/KNOW_2020.csv')

C:\Users\user\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3444: DtypeWarning: Columns (179) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


In [176]:
know_2020

idx  saq1_1 saq1_2  saq2_1 saq2_2  saq3_1 saq3_2  saq4_1 saq4_2  \
0     27127       4      4       4      4       4      5       4      5   
1     27128       5      6       5      6       4      5       4      5   
2     27129       3      4       3      4       3      4       3      5   
3     27130       4      5       2      3       3      3       4      5   
4     27131       5      6       4      5       3      5       4      4   
...     ...     ...    ...     ...    ...     ...    ...     ...    ...   
8117  35244       4      6       3      5       3      3       1          
8118  35245       3      4       4      5       2      4       4      6   
8119  35246       5      6       3      4       3      4       4      5   
8120  35247       5      6       5      5       4      5       3      3   
8121  35248       2      5       4      6       3      5       4      5   

      saq5_1  ... bq26  bq26_1 bq27_1  bq27_2 bq28  bq29 bq30_1  bq30_2  \
0          3  ...    4     간호학      1       1    1    40   4000    2600   
1          4  ...    4    기계공학      1       1    1    40   4000    3000   
2          3  ...    4      기계      1       1    1    48   4300    3000   
3          3  ...    4     보건학      1       1    1    40   3500    3300   
4          4  ...    4  건축설비공학      1       1    1    40   5700    2700   
...      ...  ...  ...     ...    ...     ...  ...   ...    ...     ...   
8117       5  ...    6    지구과학      1       1    1    40   5100    4000   
8118       1  ...    4     광고학      1       1    1    45   3100    2600   
8119       5  ...    4      국문      1       1    1    40   8000    3000   
8120       3  ...    5    화학공학      1       1    1    40   5700    3200   
8121       3  ...    3     조리학      1       1    1    46   3700    2700   

     bq30_3  knowcode  
0              304003  
1              221401  
2              815001  
3              304003  
4              140201  
...     ...       ...  
8117           121105  
8118            26102  
8119            31301  
8120           154105  
8121            14201  

[8122 rows x 185 columns]

## 이상치 파악

In [177]:
know_2020.info(verbose=1)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8122 entries, 0 to 8121
Data columns (total 185 columns):
 #    Column    Dtype 
---   ------    ----- 
 0    idx       int64 
 1    saq1_1    int64 
 2    saq1_2    object
 3    saq2_1    int64 
 4    saq2_2    object
 5    saq3_1    int64 
 6    saq3_2    object
 7    saq4_1    int64 
 8    saq4_2    object
 9    saq5_1    int64 
 10   saq5_2    object
 11   saq6_1    int64 
 12   saq6_2    object
 13   saq7_1    int64 
 14   saq7_2    object
 15   saq8_1    int64 
 16   saq8_2    object
 17   saq9_1    int64 
 18   saq9_2    object
 19   saq10_1   int64 
 20   saq10_2   object
 21   saq11_1   int64 
 22   saq11_2   object
 23   saq12_1   int64 
 24   saq12_2   object
 25   saq13_1   int64 
 26   saq13_2   object
 27   saq14_1   int64 
 28   saq14_2   object
 29   saq15_1   int64 
 30   saq15_2   object
 31   saq16_1   int64 
 32   saq16_2   object
 33   saq17_1   int64 
 34   saq17_2   object
 35   saq18_1   int64 
 36   saq18_2   ob

In [178]:
know_2020[know_2020.bq18_10.isnull()==True].index

Int64Index([235], dtype='int64')

In [179]:
know_2020.isnull().sum()[150:]

bq18_4      0
bq18_5      0
bq18_6      0
bq18_7      0
bq18_8      0
bq18_9      0
bq18_10     1
bq19        0
bq20        0
bq20_1      0
bq21_1      0
bq21_2      0
bq21_3      0
bq21_4      0
bq22_1      0
bq22_2      0
bq22_3      0
bq22_4      0
bq22_5      0
bq22_6      0
bq23_1      0
bq23_2      0
bq23_3      0
bq24        0
bq25        0
bq26        0
bq26_1      0
bq27_1      0
bq27_2      0
bq28        0
bq29        0
bq30_1      0
bq30_2      0
bq30_3      0
knowcode    0
dtype: int64

In [180]:
for i in know_2020.columns : 
    print(i, len(know_2020[i].value_counts()))

idx 8122
saq1_1 5
saq1_2 8
saq2_1 5
saq2_2 8
saq3_1 5
saq3_2 8
saq4_1 5
saq4_2 8
saq5_1 5
saq5_2 8
saq6_1 5
saq6_2 8
saq7_1 5
saq7_2 8
saq8_1 5
saq8_2 8
saq9_1 5
saq9_2 8
saq10_1 5
saq10_2 8
saq11_1 5
saq11_2 8
saq12_1 5
saq12_2 8
saq13_1 5
saq13_2 8
saq14_1 5
saq14_2 8
saq15_1 5
saq15_2 8
saq16_1 5
saq16_2 8
saq17_1 5
saq17_2 8
saq18_1 5
saq18_2 8
saq19_1 5
saq19_2 8
saq20_1 5
saq20_2 8
saq21_1 5
saq21_2 8
saq22_1 5
saq22_2 8
saq23_1 5
saq23_2 8
saq24_1 5
saq24_2 8
saq25_1 5
saq25_2 8
saq26_1 5
saq26_2 8
saq27_1 5
saq27_2 8
saq28_1 5
saq28_2 8
saq29_1 5
saq29_2 8
saq30_1 5
saq30_2 8
saq31_1 5
saq31_2 8
saq32_1 5
saq32_2 8
saq33_1 5
saq33_2 8
saq34_1 5
saq34_2 8
saq35_1 5
saq35_2 8
saq36_1 5
saq36_2 8
saq37_1 5
saq37_2 8
saq38_1 5
saq38_2 8
saq39_1 5
saq39_2 8
saq40_1 5
saq40_2 8
saq41_1 5
saq41_2 8
saq42_1 5
saq42_2 8
saq43_1 5
saq43_2 8
saq44_1 5
saq44_2 8
vq1 5
vq2 5
vq3 5
vq4 5
vq5 5
vq6 5
vq7 5
vq8 5
vq9 5
vq10 5
vq11 5
vq12 5
vq13 5
bq1 21
bq2 5
bq3 6
bq4 2
bq4_1a 1632
bq4_1b 629

In [181]:
know_2020['bq28'].value_counts()

1    3685
1    3666
2     409
2     360
        2
Name: bq28, dtype: int64

1. 결측치 

bq18_10     1

2. 이상치

bq2 근무업체규모 5 / 기본값 6

bq28 5 type 

3. 주관식 문항

bq4_1a 자격증 1632
bq4_1b 자격증 629
bq4_1c 자격증 182

bq5_2 훈련사항 1745
bq18_10 직업세계변화 1314
bq20_1 일자리변화6947

bq26_1 학과 1088

4. 숫자여야하는 col
bq25 나이 60
bq29 근로시간 58
bq30_1 근로소득 329
bq30_2 근로소득 208
bq30_3 근로소득 102

5. 체크해봐야하는 col

bq22_1 9
bq22_2 9
bq22_3 9
bq22_4 9
bq22_5 9
bq22_6 9
bq23_1 9
bq23_2 9
bq23_3 9



bq27_1 2
bq27_2 7

## 이상치 처리

In [182]:
for col in know_2020:
        know_2020[col].replace(' ', 0, inplace = True)

In [183]:
know_2020.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8122 entries, 0 to 8121
Data columns (total 185 columns):
 #    Column    Dtype 
---   ------    ----- 
 0    idx       int64 
 1    saq1_1    int64 
 2    saq1_2    object
 3    saq2_1    int64 
 4    saq2_2    object
 5    saq3_1    int64 
 6    saq3_2    object
 7    saq4_1    int64 
 8    saq4_2    object
 9    saq5_1    int64 
 10   saq5_2    object
 11   saq6_1    int64 
 12   saq6_2    object
 13   saq7_1    int64 
 14   saq7_2    object
 15   saq8_1    int64 
 16   saq8_2    object
 17   saq9_1    int64 
 18   saq9_2    object
 19   saq10_1   int64 
 20   saq10_2   object
 21   saq11_1   int64 
 22   saq11_2   object
 23   saq12_1   int64 
 24   saq12_2   object
 25   saq13_1   int64 
 26   saq13_2   object
 27   saq14_1   int64 
 28   saq14_2   object
 29   saq15_1   int64 
 30   saq15_2   object
 31   saq16_1   int64 
 32   saq16_2   object
 33   saq17_1   int64 
 34   saq17_2   object
 35   saq18_1   int64 
 36   saq18_2   ob

In [184]:
# 주관식문항 

str_li =['bq4_1a','bq4_1b','bq4_1c','bq5_2', 'bq18_10','bq20_1','bq26_1']

In [185]:
for i in know_2020.columns : 
    if i not in str_li : 
        know_2020[i] = know_2020[i].astype(int)
    else : 
        know_2020[i] = know_2020[i].astype(str)

In [186]:
know_2020.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8122 entries, 0 to 8121
Data columns (total 185 columns):
 #    Column    Dtype 
---   ------    ----- 
 0    idx       int32 
 1    saq1_1    int32 
 2    saq1_2    int32 
 3    saq2_1    int32 
 4    saq2_2    int32 
 5    saq3_1    int32 
 6    saq3_2    int32 
 7    saq4_1    int32 
 8    saq4_2    int32 
 9    saq5_1    int32 
 10   saq5_2    int32 
 11   saq6_1    int32 
 12   saq6_2    int32 
 13   saq7_1    int32 
 14   saq7_2    int32 
 15   saq8_1    int32 
 16   saq8_2    int32 
 17   saq9_1    int32 
 18   saq9_2    int32 
 19   saq10_1   int32 
 20   saq10_2   int32 
 21   saq11_1   int32 
 22   saq11_2   int32 
 23   saq12_1   int32 
 24   saq12_2   int32 
 25   saq13_1   int32 
 26   saq13_2   int32 
 27   saq14_1   int32 
 28   saq14_2   int32 
 29   saq15_1   int32 
 30   saq15_2   int32 
 31   saq16_1   int32 
 32   saq16_2   int32 
 33   saq17_1   int32 
 34   saq17_2   int32 
 35   saq18_1   int32 
 36   saq18_2   in

In [187]:
## _1에서 1을 선택한 사람중 _2를 답변한 사람이 있는지 확인

for i in range(44) :
    str5 = 'saq' + str(i+1) +'_1'
    str6 = 'saq' + str(i+1) +'_2'
    
    print(know_2020[(know_2020[str5]==1) & (know_2020[str6]!=0)][str6])
    

Series([], Name: saq1_2, dtype: int32)
Series([], Name: saq2_2, dtype: int32)
Series([], Name: saq3_2, dtype: int32)
Series([], Name: saq4_2, dtype: int32)
Series([], Name: saq5_2, dtype: int32)
Series([], Name: saq6_2, dtype: int32)
Series([], Name: saq7_2, dtype: int32)
Series([], Name: saq8_2, dtype: int32)
Series([], Name: saq9_2, dtype: int32)
Series([], Name: saq10_2, dtype: int32)
Series([], Name: saq11_2, dtype: int32)
Series([], Name: saq12_2, dtype: int32)
Series([], Name: saq13_2, dtype: int32)
Series([], Name: saq14_2, dtype: int32)
Series([], Name: saq15_2, dtype: int32)
Series([], Name: saq16_2, dtype: int32)
Series([], Name: saq17_2, dtype: int32)
Series([], Name: saq18_2, dtype: int32)
Series([], Name: saq19_2, dtype: int32)
Series([], Name: saq20_2, dtype: int32)
Series([], Name: saq21_2, dtype: int32)
Series([], Name: saq22_2, dtype: int32)
Series([], Name: saq23_2, dtype: int32)
Series([], Name: saq24_2, dtype: int32)
Series([], Name: saq25_2, dtype: int32)
Series([]

In [188]:
know_2020.iloc[2177, 80] = 0

In [189]:
## 결측치 파악(답변있어야하는 셀 )

for i in range(44) :
    str5 = 'saq' + str(i+1) +'_1'
    str6 = 'saq' + str(i+1) +'_2'
    
    print(know_2020[(know_2020[str5].isin([2,3,4,5])) & (know_2020[str6]==0)][str6])
    
#선택질문 중 빈셀 없음 

Series([], Name: saq1_2, dtype: int32)
Series([], Name: saq2_2, dtype: int32)
Series([], Name: saq3_2, dtype: int32)
Series([], Name: saq4_2, dtype: int32)
Series([], Name: saq5_2, dtype: int32)
Series([], Name: saq6_2, dtype: int32)
Series([], Name: saq7_2, dtype: int32)
Series([], Name: saq8_2, dtype: int32)
Series([], Name: saq9_2, dtype: int32)
Series([], Name: saq10_2, dtype: int32)
Series([], Name: saq11_2, dtype: int32)
Series([], Name: saq12_2, dtype: int32)
Series([], Name: saq13_2, dtype: int32)
Series([], Name: saq14_2, dtype: int32)
Series([], Name: saq15_2, dtype: int32)
Series([], Name: saq16_2, dtype: int32)
Series([], Name: saq17_2, dtype: int32)
Series([], Name: saq18_2, dtype: int32)
Series([], Name: saq19_2, dtype: int32)
Series([], Name: saq20_2, dtype: int32)
Series([], Name: saq21_2, dtype: int32)
Series([], Name: saq22_2, dtype: int32)
Series([], Name: saq23_2, dtype: int32)
Series([], Name: saq24_2, dtype: int32)
Series([], Name: saq25_2, dtype: int32)
Series([]

1. 주관식 항목 drop예정으로 확인 안함 

bq18_10 bq20_1

2. 학과
bq26_1 빈셀 없음

3. bq27_1 ~ bq30_3
bq28,29 는 조건에 맞게 대답하려면 빈칸이 많아야하는데 그렇지 않음 그냥 빈셀을 채워주는 식으로 처리 할 예정 

bq28 빈셀 2개 
28965(1838) -> 2 
30547(3420) -> 1


bq29 빈셀 없음

2017과 동일하게 처리 bq30_1 빈셀 1개 bq30_2 빈셀 3개 bq30_3 빈셀 없음

In [190]:
know_2020.iloc[1838,179]=2

In [191]:
know_2020.iloc[3420,179]=1

In [192]:
null_li = know_2020[(know_2020.bq27_1==1) & (know_2020.bq30_1==0)].index
null_li

Int64Index([1854], dtype='int64')

In [193]:
for i in null_li:
      know_2020.iloc[i,181]= round(know_2020[(know_2020.bq1 ==know_2020.iloc[i,102])&(know_2020.bq30_1!=0)].bq30_1.mean())

In [194]:
null_li = know_2020[(know_2020.bq27_1==1) & (know_2020.bq30_1==0)].index
null_li

Int64Index([], dtype='int64')

In [195]:
null_li2 = know_2020[(know_2020.bq27_1==1) & (know_2020.bq30_2==0)].index
null_li2

Int64Index([4172, 4882, 7430], dtype='int64')

In [196]:
for i in null_li2:
      know_2020.iloc[i,182]= round(know_2020[(know_2020.bq1 ==know_2020.iloc[i,102])&(know_2020.bq30_2!=0)].bq30_2.mean())

In [197]:
null_li2 = know_2020[(know_2020.bq27_1==1) & (know_2020.bq30_2==0)].index
null_li2

Int64Index([], dtype='int64')

In [198]:
## 자격증 없음 

know_2020['bq4_1b'].replace(0, '0', inplace = True)


for i in know_2020.bq4_1b:
    if '없' in i : 
        print(i)

know_2020['bq4_1b'].replace('없다', '0', inplace = True)

없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다


In [199]:
for i in know_2020.bq4_1b:
    if '없' in i : 
        print(i)

In [200]:
know_2020.to_csv("know_2020_enutf.csv", encoding='utf8', index=False)
know_2020.to_csv("know_2020_encp.csv", encoding='cp949', index=False)

In [201]:
df_cp = pd.read_csv("know_2020_encp.csv", encoding='cp949')
df_cp.info(verbose=1)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8122 entries, 0 to 8121
Data columns (total 185 columns):
 #    Column    Dtype 
---   ------    ----- 
 0    idx       int64 
 1    saq1_1    int64 
 2    saq1_2    int64 
 3    saq2_1    int64 
 4    saq2_2    int64 
 5    saq3_1    int64 
 6    saq3_2    int64 
 7    saq4_1    int64 
 8    saq4_2    int64 
 9    saq5_1    int64 
 10   saq5_2    int64 
 11   saq6_1    int64 
 12   saq6_2    int64 
 13   saq7_1    int64 
 14   saq7_2    int64 
 15   saq8_1    int64 
 16   saq8_2    int64 
 17   saq9_1    int64 
 18   saq9_2    int64 
 19   saq10_1   int64 
 20   saq10_2   int64 
 21   saq11_1   int64 
 22   saq11_2   int64 
 23   saq12_1   int64 
 24   saq12_2   int64 
 25   saq13_1   int64 
 26   saq13_2   int64 
 27   saq14_1   int64 
 28   saq14_2   int64 
 29   saq15_1   int64 
 30   saq15_2   int64 
 31   saq16_1   int64 
 32   saq16_2   int64 
 33   saq17_1   int64 
 34   saq17_2   int64 
 35   saq18_1   int64 
 36   saq18_2   in

In [202]:
df_utf = pd.read_csv("know_2020_enutf.csv")
df_utf.info(verbose=1)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8122 entries, 0 to 8121
Data columns (total 185 columns):
 #    Column    Dtype 
---   ------    ----- 
 0    idx       int64 
 1    saq1_1    int64 
 2    saq1_2    int64 
 3    saq2_1    int64 
 4    saq2_2    int64 
 5    saq3_1    int64 
 6    saq3_2    int64 
 7    saq4_1    int64 
 8    saq4_2    int64 
 9    saq5_1    int64 
 10   saq5_2    int64 
 11   saq6_1    int64 
 12   saq6_2    int64 
 13   saq7_1    int64 
 14   saq7_2    int64 
 15   saq8_1    int64 
 16   saq8_2    int64 
 17   saq9_1    int64 
 18   saq9_2    int64 
 19   saq10_1   int64 
 20   saq10_2   int64 
 21   saq11_1   int64 
 22   saq11_2   int64 
 23   saq12_1   int64 
 24   saq12_2   int64 
 25   saq13_1   int64 
 26   saq13_2   int64 
 27   saq14_1   int64 
 28   saq14_2   int64 
 29   saq15_1   int64 
 30   saq15_2   int64 
 31   saq16_1   int64 
 32   saq16_2   int64 
 33   saq17_1   int64 
 34   saq17_2   int64 
 35   saq18_1   int64 
 36   saq18_2   in